File: python/flood_fill.py
from typing import List

class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        r_lim, c_lim, orig = len(image), len(image[0]), image[sr][sc]
        def dfs(r, c):
            if r < 0 or r >= r_lim or c < 0 or c >= c_lim or image[r][c] != orig:
                return
            image[r][c] = color
            for dx, dy in ((-1, 0), (0, -1), (0, 1), (1, 0)):
                new_r, new_c = r + dx, c + dy
                dfs(new_r, new_c)
        if orig != color: #if it already has the new color, there is no need to call dfs
            dfs(sr, sc)
        return image

File: cpp/flood_fill.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution
{
public:
    void dfs(vector<vector<int>> &image, int r, int c, int orig, int color)
    {
        if (r < 0 || r >= image.size() || c < 0 || c >= image[0].size() || image[r][c] != orig)
        {
            return;
        }

        image[r][c] = color;
        int d[4][2] = {{-1, 0}, {0, -1}, {0, 1}, {1, 0}};
        for (int i = 0; i < 4; i++)
        {
            int nr = r + d[i][0], nc = c + d[i][1];
            dfs(image, nr, nc, orig, color);
        }
    }

    vector<vector<int>> floodFill(vector<vector<int>> &image, int sr, int sc, int color)
    {
        if (image[sr][sc] != color)
        {
            dfs(image, sr, sc, image[sr][sc], color);
        }
        return image;
    }
};